src/runtime_src/core/common/drv/cu_scu.c

```diff
@@ -0,0 +1,211 @@
+/* SPDX-License-Identifier: GPL-2.0 OR Apache-2.0 */
+/*
+ * Xilinx Soft CU
+ *
+ * Copyright (C) 2022 Xilinx, Inc. All rights reserved.
+ *
+ * Authors: min.ma@xilinx.com, j.lin@xilinx.com
+ *
+ * This file is dual-licensed; you may select either the GNU General Public
+ * License version 2 or Apache License, Version 2.0.
+ */
+
+#include "xrt_cu.h"
+#include "xgq_cmd_ert.h"
+#include "zocl_drv.h"
+
+/* Control register bits and special behavior if any.
+ * Bit 0: ap_start(Read/Set). Clear by CU when ap_ready assert.
+ * Bit 1: ap_done(Read only). Clear on read.
+ * Bit 2: ap_idle(Read only).
+ * Bit 3: ap_ready(Read only). Self clear after clear ap_start.
+ * Bit 4: ap_continue(Read/Set). Self clear.
+ * Bit 5-7: Not support yet
+ */
+#define CTRL		0x0
+#define ARGS		0x4
+
+extern int kds_echo;
+
+static int scu_alloc_credit(void *core)
+{
+	struct xrt_cu_scu *scu = (struct xrt_cu_scu *)core;
+
+	return (scu->credits) ? scu->credits-- : 0;
+}
+
+static void scu_free_credit(void *core, u32 count)
+{
+	struct xrt_cu_scu *scu = (struct xrt_cu_scu *)core;
+
+	scu->credits += count;
+	if (scu->credits > scu->max_credits)
+		scu->credits = scu->max_credits;
+}
+
+static int scu_peek_credit(void *core)
+{
+	struct xrt_cu_scu *scu = (struct xrt_cu_scu *)core;
+
+	return scu->credits;
+}
+
+static void scu_xgq_start(struct xrt_cu_scu *scu, u32 *data)
+{
+	struct xgq_cmd_start_cuidx *cmd = (struct xgq_cmd_start_cuidx *)data;
+	u32 num_reg = 0;
+	u32 i = 0;
+	u32 *cu_regfile;
+
+	cu_regfile = scu->vaddr;
+	num_reg = (cmd->hdr.count - (sizeof(struct xgq_cmd_start_cuidx)
```

Comment:
For 1), cu_scu get "data" which is pointing to a XGQ command. We need to calculate the payload here, which is for PS kernel thread. We cannot know the details of the XGQ command before it goes into scu module.
For 2), You are right. The cu_regfile[] array is 4KB. This is an assumption in the code. This can be done in a data driven way.
For 3), The data size is the size of the XGQ command. It should larger than the num_reg.